### DISCUSSION TITLE
Weird TPM result numbers for customscript in CLI

### DISCUSSION THREAD

**Discussion by xychar** [TPC-Council/HammerDB#70] [Q&A] [answered]
Steps to reproduce (using any databases, like mssqls or postgres)

Open hammerdbcli.bat

1.  Set the parameters, like:
    1. dbset db pg
    2. dbset bm TPC-C
    3. set other parameters
    4. set driver mode - timed
2. Run loadscript, and save it to a file
3. Run customscript, load it from a file
4. Run the benchmark
    1. vuset vu 4
    2. vucreate
    3. vurun
    4. wait it to finish
5. You will get very small number for the TPM and NOPM
6. If you run it once more, will get reasonable results

**Comment by sm-shaw** [comment]
A few things here. Firstly the code that runs the workload in the CLI is exactly the same as the code in the GUI (this part of the code is shared) - so one troubleshooting exercise is to repeat the finding with the GUI. Secondly you are loading a customscript so a lot does depend on what is in the customscript and how much it differs from the standard scripts that are loaded. Thirdly - depending on your I/O configuration and the rampup time and the buffer cache size the workload will spend a lot of time initially reading data from disk and loading it into the cache - you can see this from your database statistics - therefore if you watch the first workload while running the transaction counter you will see a ramp up in the graph of the workload. Running the transaction counter is good practice for troubleshooting anyway as you are aiming for a flat graph at peak transaction rates.

**Comment by xychar** [comment]
Okey, here's the detailed steps.

Firstly, let's have a look at what we get from running "loadscript":
#Using hammdbcli

dbset db mssqls
dbset bm TPC-C

diset connection mssqls_tcp true
diset connection mssqls_authentication sql

diset connection mssqls_server test-server-01
diset connection mssqls_port 1433
diset connection mssqls_uid master
diset connection mssqls_pass Test#W0rd

diset tpcc mssqls_count_ware 16
diset tpcc mssqls_num_vu 8
diset tpcc mssqls_dbase tpcc_tpm

vudestroy

diset tpcc mssqls_driver timed
diset tpcc mssqls_rampup 1
diset tpcc mssqls_duration 2
diset tpcc mssqls_allwarehouse true

% loadscript
Script loaded, Type "print script" to view

#save it to c1.tcl
print script


% vuset vu 4
% vucreate
Vuser 1 created MONITOR - WAIT IDLE
Vuser 2 created - WAIT IDLE
Vuser 3 created - WAIT IDLE
Vuser 4 created - WAIT IDLE
Vuser 5 created - WAIT IDLE
5 Virtual Users Created with Monitor VU
% vustatus
1 = WAIT IDLE
2 = WAIT IDLE
3 = WAIT IDLE
4 = WAIT IDLE
5 = WAIT IDLE

We set the virtual-user number as 4,
but, we can see there 5 virtual users are created, one is for monitoring?

**Comment by xychar** [comment]
Then, let's see how "customscript" run:
Need restart the hammerdbcli to get a fresh environment.

dbset db mssqls
dbset bm TPC-C

diset connection mssqls_tcp true
diset connection mssqls_authentication sql

diset connection mssqls_server test-server-01
diset connection mssqls_port 1433
diset connection mssqls_uid master
diset connection mssqls_pass Test#W0rd

diset tpcc mssqls_count_ware 16
diset tpcc mssqls_num_vu 8
diset tpcc mssqls_dbase tpcc_tpm

vudestroy

diset tpcc mssqls_driver timed
diset tpcc mssqls_rampup 1
diset tpcc mssqls_duration 2
diset tpcc mssqls_allwarehouse true

% customscript c1.tcl
Loaded c1.tcl

% vuset vu 4
% vucreate
Vuser 1 created - WAIT IDLE
Vuser 2 created - WAIT IDLE
Vuser 3 created - WAIT IDLE
Vuser 4 created - WAIT IDLE
4 Virtual Users Created
% vustatus
1 = WAIT IDLE
2 = WAIT IDLE
3 = WAIT IDLE
4 = WAIT IDLE


Now, we can see, even we set the vu as 4, but only 4 virtual users are created.
One of them is the monitoring virtual user, so we actually get only 3 virtual users.

The final benchmark results also confirm that there are only 3 virtual users.

**Comment by xychar** [comment]
That's the difference between loadscript and customscript, is that expected?

**Comment by xychar** [comment]
For customscript:

% vurun
RUNNING -
Vuser 1:RUNNING
Vuser 1:Beginning rampup time of 1 minutes
Vuser 2:RUNNING
Vuser 2:Processing 1000000 transactions with output suppressed...
Vuser 3:RUNNING
Vuser 3:Processing 1000000 transactions with output suppressed...
Vuser 4:RUNNING
Vuser 4:Processing 1000000 transactions with output suppressed...
% vustatus
1 = RUNNING
2 = RUNNING
3 = RUNNING
4 = RUNNING

**Comment by xychar** [comment]
The full log:

hammerdb>customscript c1.tcl
Loaded c1.tcl

hammerdb>vuset vu 4

hammerdb>vuset logtotemp 0

hammerdb>vucreate
Vuser 1 created - WAIT IDLE
Vuser 2 created - WAIT IDLE
Vuser 3 created - WAIT IDLE
Vuser 4 created - WAIT IDLE
4 Virtual Users Created

hammerdb>vurun
RUNNING -
Vuser 1:RUNNING
Vuser 1:Beginning rampup time of 1 minutes
Vuser 2:RUNNING
Vuser 2:Processing 1000000 transactions with output suppressed...
Vuser 3:RUNNING
Vuser 3:Processing 1000000 transactions with output suppressed...
Vuser 4:RUNNING
Vuser 4:Processing 1000000 transactions with output suppressed...

hammerdb>


hammerdb>

Vuser 1:Rampup 1 minutes complete ...
Vuser 1:Rampup complete, Taking start Transaction Count.
Vuser 1:Timing test period of 2 in minutes
Vuser 1:1 ...,
Vuser 1:2 ...,
Vuser 1:Test complete, Taking end Transaction Count.
Vuser 1:3 Active Virtual Users configured
Vuser 1:TEST RESULT : System achieved 199434 SQL Server TPM at 43378 NOPM
Vuser 1:FINISHED SUCCESS
Vuser 2:FINISHED SUCCESS
Vuser 4:FINISHED SUCCESS
Vuser 3:FINISHED SUCCESS
ALL VIRTUAL USERS COMPLETE

**Comment by sm-shaw** [comment, marked as answer]
Yes the difference between loadscript for a timed workload and customscript is expected and intentional.  When a timed script is loaded it creates an additional virtual user for monitoring. The best way to observe this behaviour is using the GUI.  Then select a test script and load it - you can see that there is no additional monitoring virtual user. If you choose a customscript this is the same as selecting file load from the menu. HammerDB cannot know what your script contains so it does not create the monitor. This is all expected behaviour. As background the reason for this functionality was that it was observed that testers beginning with HammerDB would often make the mistake of running a timed script with 1 virtual user and therefore ran a monitor only without any active virtual users running the workload. Therefore the default behaviour was changed to create the monitor automatically for a timed workload. As noted previously it is recommended to use the GUI to see how this operates.

**Comment by sm-shaw** [comment]
Issue closed as behaviour is as expected

**Comment by xychar** [comment]
It obviously does not make sense.

hammerdb>vuset vu 4

but, finally got:

Vuser 1:3 Active Virtual Users configured
Vuser 1:TEST RESULT : System achieved 199434 SQL Server TPM at 43378 NOPM

**Comment by sm-shaw** [comment]
Please use the GUI to get familiar with how HammerDB works and then move to the CLI. You should also be looking for larger warehouse counts and longer rampup and test times for correct testing. The TPM and NOPM can go into multiple millions of transactions so your tests are at the very low end of performance for now.

**Comment by xychar** [comment]
I mean this inconsistency:

vuset vu 4

but
3 Active Virtual Users configured

**Comment by xychar** [comment]
Sorry, I think it is my fault that I did not express clearly.

**Comment by xychar** [comment]
When I test with 1 virtual user:
vuset vu 1,

for the customscript, will get:

0 Active Virtual Users configured
TEST RESULT : System achieved 8 Oracle TPM at 0 NOPM

That's what I said, the weird number of TPM.

**Comment by sm-shaw** [comment]
Yes correct this behaviour is exactly as described above - there are 0 Active Virtual Users so only the monitor virtual user is running so there are no transactions.  Clearscript will reset the status before loadscript.
